src/demo/spawn.ts

```diff
@@ -0,0 +1,283 @@
+            provider: options.provider || 'local'
+        };
+
+        const manifest = `http://localhost:3922/manifest?${
+            Object.keys(queryOptions)
+                .map(key => {
+                    return `${key}=${encodeURIComponent(queryOptions[key].toString())}`;
+                })
+                .join('&')}`;
+
+        return startApp(fin.Application.createFromManifest(manifest));
+    }
+}
+
+async function createChildWindow(data: Omit<WindowData, 'parent'>): Promise<_Window> {
+    const name: string = data.id || `Win-${Math.random().toString().substr(2, 4)}`;
+    const url = getUrl(data);
+    const position = getWindowPosition(data);
+    const size = getWindowSize(data);
+
+    const options:
+        WindowOption = {...position, name, url, frame: data.frame, autoShow: true, saveWindowState: false, defaultWidth: size.x, defaultHeight: size.y};
+    return fin.Window.create(options);
+}
+
+async function startApp(appPromise: Promise<Application>): Promise<Application> {
+    const app = await appPromise;
+    await app.run();
+    return app;
+}
+
+function getUrl(options: WindowData): string {
+    // Create URL from base URL + queryArgs
+    let url = options.url || 'http://localhost:3922/test/test-app.html';
+    const urlQueryParams = options.queryArgs || {};
+    const urlQueryKeys = Object.keys(urlQueryParams);
+
+    // Resolve relative URL's
+    if (url.indexOf('://') === -1) {
+        // No protocol, assume relative URL and resolve against location.href
+        url = new URL(url, location.href).href;
+    }
+
+    // Add-on querystring arguments
+    const queryParams: string[] = [];
+    Object.keys(urlQueryParams).forEach(param => {
+        const value = urlQueryParams[param];
+        if (value) {
+            queryParams.push(`${param}=${encodeURIComponent(typeof value === 'object' ? JSON.stringify(value) : value.toString())}`);
+        }
+    });
+    if (urlQueryParams && urlQueryKeys.length > 0) {
+        url += `${url.includes('?') ? '&' : '?'}${queryParams.join('&')}`;
+    }
+
+    return url;
+}
+
+function getWindowPosition(options: WindowData): {defaultCentered?: boolean; defaultLeft?: number; defaultTop?: number} {
+    const {position} = options;
+
+    if (position === 'center') {
+        // Position in center of screen
+        return {defaultCentered: true};
+    } else if (position) {
+        // Position in fixed x/y position
+        const out: {defaultLeft?: number; defaultTop?: number} = {};
+        if (position.x !== undefined) {
+            out.defaultLeft = position.x;
+        }
+        if (position.y !== undefined) {
+            out.defaultTop = position.y;
+        }
+        return out;
+    } else {
+        // Use system-default positioning
+        return {};
+    }
+}
+
+function getWindowSize(options: WindowData): Point {
+    return {
+        x: (options.size && options.size.x || 400) * (1 + (Math.random() * (options.sizeOffset || 0))),
```

Comment:
Oh wow, I'd missed this till now. We have Math.random() in a base utility for our layouts tests?!

Reply:
Apparently? I didn't really look at this much.

Reply:
Only if you use `sizeOffset`. Handy in the layouts demo app for manually testing snapping of slightly-differently-sized windows. But yeah, probably best to avoid in the automated tests.

Reply:
We're unlikely to specify a sizeOffset anyway, so would you rather I remove it or just leave it in to make it easier to extract them all later (since layouts will rely on it)